### PR TITLE
shuffle2 foundation

### PR DESCRIPTION
**Pull request by DenisYaroshevskiy**
Ready to be reviewed.

Everything about shuffle_v2 is it's own folder and test folder too. I don't think we want this swissle/shuffle division.

Canonical shuffle is now shuffle_v2_driver. It's a component that takes the input from the user and simplifies/splits it enough to the go into native shuffles.

native_shuffle selector is the thing that calls l2, l3 ... shuffles one after enother.
The comments in `shuffle_v2_fwd` explain the rational for levels.

The `_fwd` are extracted because we want to recurse into shuffles with previous levels.

Previous `shuffle_l0` tests are migrated to legacy_l0_tests.
We want tests that describe the pattern and then match that to the appropirate level.

### CODE REVIEW

File: include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+  else return shuffle_v2_driver_bundle {}(selector, p, g, xs);
+}
+
+// entry point -----------------------------------------
+
+template<typename T, typename... Ts>
+EVE_FORCEINLINE auto
+shuffle_v2_driver_construct_result(kumi::tuple<T, Ts...> shuffled)
+{
+  if constexpr( sizeof...(Ts) == 0 ) return get<0>(shuffled);
```
Comment (jfalcou) [inline include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp:578]: this is quite a large file, any hope to split it up in smaller chunks ?

File: include/eve/detail/shuffle_v2/idxm.hpp [hunk 2]
```diff
@@ -211,4 +224,177 @@ is_rotate(std::span<const std::ptrdiff_t> idxs)
...
+      if( register_used[i] ) { *out++ = (int)i; }
+    }
+
+    // no registers means one register.
+    if( out == r.data() ) ++out;
+
+    return std::pair {r, (std::size_t)(out - r.data())};
+  }();
+
+  constexpr auto used_wide_indexes = [&]
```
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply include/eve/detail/shuffle_v2/idxm.hpp:284]: Oh this one is super nasty.
Basically - I can't allocate (not in our compilers yet unfortunately).

So I need to figure out array sizes.

File: include/eve/detail/shuffle_v2/idxm.hpp [hunk 3]
```diff
@@ -211,4 +224,177 @@ is_rotate(std::span<const std::ptrdiff_t> idxs)
...
+    if( idxs[i] == we_ || pattern_f[i] == we_ ) continue;
+    if( idxs[i] != pattern_f[i] ) return false;
+  }
+
+  return true;
+}
+
+// Can't variadic this
+constexpr bool
+matches(const std::span<const std::ptrdiff_t> idxs,
```
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply include/eve/detail/shuffle_v2/idxm.hpp:333]: That doesn't work unfortunately. https://godbolt.org/z/cEnGe98M9

File: include/eve/detail/shuffle_v2/idxm.hpp [hunk 4]
```diff
@@ -211,4 +224,177 @@ is_rotate(std::span<const std::ptrdiff_t> idxs)
...
+  return res;
+}
+
+template<std::ptrdiff_t G, std::size_t N>
+constexpr std::array<std::ptrdiff_t, N * G>
+expand_group(std::array<std::ptrdiff_t, N> idxs)
+{
+  return expand_group<G>(std::span<const std::ptrdiff_t, N>(idxs));
+}
+
```
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply include/eve/detail/shuffle_v2/idxm.hpp:395]: This is just a namespace to do operations on indexes. 

Just "is this a rotation", "is this a repeated pattern" etc.

File: include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+{
+  constexpr std::array    p_arr {I...};
+  constexpr std::optional attempt = upscale_pattern_impl(p_arr);
+  if constexpr( !attempt ) return p;
+  else return idxm::to_pattern<*attempt>();
+}
+
+template<typename U, typename... Ts>
+EVE_FORCEINLINE auto
+bit_cast_tuple(kumi::tuple<Ts...> xs, eve::as<U>)
```
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp:117]: I think that's way too cute. I just need it for this specific problem, it's in the depth of detail.

File: include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp:578]: I can move `simplify_plain_shuffle` into a separate file. Let me do that in a separate patch if you don't mind.

File: include/eve/detail/shuffle_v2/idxm.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/detail/shuffle_v2/idxm.hpp:333]: Too bad :{

File: include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/detail/shuffle_v2/simd/common/shuffle_v2_driver.hpp:117]: OK

**Review by jfalcou — APPROVED** [review]

**Pull request merged**